builder/openstack/server.go

```diff
@@ -138,11 +138,6 @@ func DeleteServer(state multistep.StateBag, instance string) error {
 	}
 
 	server, err := servers.Get(computeClient, instance).Extract()
```

Comment:
I think we should also remove this.

Reply:
Line 140? That's needed for lines:
        stateChange := StateChangeConf{
                Pending: []string{"ACTIVE", "BUILD", "REBUILD", "SUSPENDED", "SHUTOFF", "STOPPED"},
                Refresh: ServerStateRefreshFunc(computeClient, server),
                Target:  []string{"DELETED"},
        }

Reply:
Ah sorry, of course! I remember now, that bit of code correctly treats 404 as success when checking for a deleted server, which the code you have removed did not.

Reply:
Hi @jjo93sa and @JohnGarbutt - @lbajolet-hashicorp and I had a chance to review to better understand what is happening here. With our understanding of the code we suggest you make one of the following two changes to better support this issue. 

1. The preferred change used in three places would be to drop the pointer to server for just the instance ID in ServerStateRefreshFunc.
```go
// ServerStateRefreshFunc returns a StateRefreshFunc that is used to watch
// an openstack server.
func ServerStateRefreshFunc(
	client *gophercloud.ServiceClient, instanceID string) StateRefreshFunc {
	return func() (interface{}, string, int, error) {
		serverNew, err := servers.Get(client, instanceID).Extract()
		if err != nil {
			if _, ok := err.(gophercloud.ErrDefault404); ok {
				log.Printf("[INFO] 404 on ServerStateRefresh, returning DELETED")
				return nil, "DELETED", 0, nil
			}
			log.Printf("[ERROR] Error on ServerStateRefresh: %s", err)
			return nil, "", 0, err
		}

		return serverNew, serverNew.Status, serverNew.Progress, nil
	}
}
```
2. Move lines 140 - 144 before the deletion loop this way you have a handle to the valid server before it gets deleted. The results of `server, err := servers.Get(computeClient, instance).Extract()` will be used by `Refresh: ServerStateRefreshFunc(computeClient, server),` on line 143. As server can't be nil, if the call to Get does error then it would be valid to error but it would because of something else not a race condition error. 

```go
	server, err := servers.Get(computeClient, instance).Extract()
	if err != nil {
		err = fmt.Errorf("Error getting server to terminate: %s", err)
		return err
	}

	ui.Say(fmt.Sprintf("Terminating the source server: %s ...", instance))
	for {
		if config.ForceDelete {
			err = servers.ForceDelete(computeClient, instance).ExtractErr()
		} else {
			err = servers.Delete(computeClient, instance).ExtractErr()
		}

		if err == nil {
			break
		}

		if _, ok := err.(gophercloud.ErrDefault500); !ok {
			err = fmt.Errorf("Error terminating server, may still be around: %s", err)
			return err
		}

		if numErrors < maxNumErrors {
			numErrors++
			log.Printf("Error terminating server on (%d) time(s): %s, retrying ...", numErrors, err)
			time.Sleep(2 * time.Second)
			continue
		}
		err = fmt.Errorf("Error terminating server, maximum number (%d) reached: %s", numErrors, err)
		return err
	}

	stateChange := StateChangeConf{
		Pending: []string{"ACTIVE", "BUILD", "REBUILD", "SUSPENDED", "SHUTOFF", "STOPPED"},
		Refresh: ServerStateRefreshFunc(computeClient, server),
		Target:  []string{"DELETED"},
	}
```